teachopencadd/talktorials/T018_automated_cadd_pipeline/utils/helpers/pdb.py

```diff
@@ -0,0 +1,173 @@
+        Local path of the original PDB file.
+    output_filepath : str or pathlib.Path
+        Local file path (including file name) to save the PDB file of the extracted molecule in.
+
+    Returns
+    -------
+    MDAnalysis.core.universe.Universe
+        Structure object of the extracted molecule.
+    """
+
+    pdb_structure = Structure.from_string(input_filepath)
+    molecule_name = f"resname {molecule_name}" if molecule_name != "protein" else molecule_name
+    extracted_structure = pdb_structure.select_atoms(molecule_name)
+    extracted_structure.write(output_filepath)
+    return extracted_structure
+
+
+def load_pdb_file_as_dataframe(pdb_file_text_content):
+    """
+    Transform the textual content of a PDB file into a dictionary of pandas DataFrames.
+
+    Parameters
+    ----------
+    pdb_file_text_content : str
+        Textual content of a PDB file as a single string.
+
+    Returns
+    -------
+    Dict of pandas.DataFrames
+        The dictionary has 4 entries with following keys: 'ATOM', 'HETATM', 'ANISOU' and 'OTHERS'.
+        Each value is a DataFrame corresponding to the specific information described by the key.
+    """
+    ppdb = PandasPdb().read_pdb_from_list(pdb_file_text_content.splitlines(True))
+    pdb_df = ppdb.df
+    return pdb_df
+
+
+def extract_info_from_pdb_file_content(pdb_file_text_content):
+    """
+    Extract some useful information from the contents of a PDB file.
+
+    Parameters
+    ----------
+    pdb_file_text_content : str
+        Textual content of a PDB file as a single string.
+
+    Returns
+    -------
+    dict
+        Dictionary of the successfully extracted information.
+        Possible keys are:
+            'structure_title' : str
+                Title of the PDB structure.
+            'name' : str
+                Name of the protein.
+            'chains' : list of str
+                List of chain-IDs of the available chains in the protein.
+            'ligands' : list of list of str
+                List of ligand information:
+                [ligand-ID, chain-ID+residue number, number of heavy atoms]
+    """
+
+    pdb_content = pdb_file_text_content.strip().split("\n")
+    for index in range(len(pdb_content)):
+        pdb_content[index] = pdb_content[index].split(" ", 1)
+        try:
+            pdb_content[index][1] = pdb_content[index][1].strip()
+            if pdb_content[index][1][-1] == ";":
+                pdb_content[index][1] = pdb_content[index][1][:-1]
+            if (
+                pdb_content[index][0] in ["TITLE", "REMARK", "COMPND"]
+                and pdb_content[index][1][0].isdigit()
+            ):
+                try:
+                    pdb_content[index][1] = pdb_content[index][1].split(" ", 1)[1]
+                except:
+                    # FIXME specify exception
+                    pass
+        except:
+            # FIXME specify exception
+            pdb_content[index].append(" ")
```

Comment:
@Armin-Ariamajd, can you please comment on each line of code here? (We will need to fix the generic `except`.)

Reply:
So this function is a (rather crude) PDB file parser that I wrote. However I noticed that not all PDB files follow the same exact format, and it is thus not very easy to parse the text of all PDB files and extract useful information using simple regex. 
The explanation of this code is a bit more involved, but I'll try. However, **in short, these try-except blocks are again not there to really catch errors, but to ignore specific lines of the PDB file that we are not interested in, and for which the code doesn't work. So there is no need really to raise an exception or anything in this case either.**

So, the logic is as follows:
The function takes in the file content of a PDB file as a single string (generated by the `pdb.read_pdb_file_content` function), which looks like this:
```
"HEADER    TRANSFERASE/TRANSFERASE INHIBITOR       07-DEC-12   3W32              \nTITLE     EGFR KINASE DOMAIN COMPLEXED WITH COMPOUND 20A                        \nCOMPND    MOL_ID: 1;                                                            \nCOMPND   2 MOLECULE: EPIDERMAL GROWTH FACTOR RECEPTOR; ...."
```
Now I'll explain line by line what is happening:
1.
```python
pdb_content = pdb_file_text_content.strip().split("\n")
```
The single string is broken up into lines (based on "\n" separators in the string), so now `pdb_content` is a list of strings, where each string is a line in the PDB file. Something like this:
```
['HEADER    TRANSFERASE/TRANSFERASE INHIBITOR       07-DEC-12   3W32              ',
 'TITLE     EGFR KINASE DOMAIN COMPLEXED WITH COMPOUND 20A                        ', ...
```
2. 
```python
for index in range(len(pdb_content)):
    pdb_content[index] = pdb_content[index].split(" ", 1)
```
Now we iterate the list, and again split every line once, at the first instance of a space. The result will be a list of lists, where each sub-list corresponds to a split line, where the first element is the information identifier (e.g. TITLE, COMPND etc.) and the second element is the information itself, like this:
```
[['HEADER',
  '   TRANSFERASE/TRANSFERASE INHIBITOR       07-DEC-12   3W32              '],
 ['TITLE',
  '    EGFR KINASE DOMAIN COMPLEXED WITH COMPOUND 20A                        '], ...
```
3. We are only interested in extracting the following information: title of the structure, name of the protein, chain-IDs, and ligand-IDs. These have the following identifiers, respectively: TITLE, COMPND (followed by MOLECULE), COMPND (followed by CHAIN), and HET. We want to match the identifier (i.e. first element of each sub-list) and then read its corresponding value (i.e. second element of the sub-list). But the problem is that not all of the lines contain spaces (there are the lines we are not interested in), therefore there would be sub-lists with only one element, for which calling the second element will raise an IndexError. But since we are not interested in those lines, we wrap the rest of the code in a `try-except` block, so that we ignore the lines with no spaces (and thus with single-element sub-lists):
```python
try:
    pdb_content[index][1] = pdb_content[index][1].strip()
    if pdb_content[index][1][-1] == ";":
        pdb_content[index][1] = pdb_content[index][1][:-1]
```
So, here we try to read the second element of each sub-list (the actual information). We then first strip the string of any pre- and post-spaces. Then we get rid of semi-colons at the end of the line.
4.
```
if (
        pdb_content[index][0] in ["TITLE", "REMARK", "COMPND"]
        and pdb_content[index][1][0].isdigit()
    ):
        try:
            pdb_content[index][1] = pdb_content[index][1].split(" ", 1)[1]
        except:
            # FIXME specify exception
            pass
```
Now, the lines starting with TITLE, REMARK, and COMPND are sometimes followed by a number (itself followed by a space). We thus check if that is the case, and if so, we remove the number with another `split` and take the second element of the resulting list. Now the problem is that some lines only have a number and nothing after it, so again, we have to wrap that inside a `try-except` block, because for those line with only a number, the list resulting from the split would not have any second element to take, and thus would raise an IndexError again.

I hope I could explain it good enough.
I'm sure there are much better ways to parse a PDB file more efficiently, but I couldn't find a function in the packages I used that can provide these information (i.e. chain-IDs, ligand-IDs etc.), so I came up with this quick and dirty way to extract these specific information from the PDB files.

Reply:
Oh, and by the way, I just saw your comment on line 160 of `pdb.py` (just after this `try-except` block I just commented about):
```python
info = {"Structure Title": [], "Name": [], "Chains": [], "Ligands": []}
    for content in pdb_content:
        # FIXME this should be elifs, no?
        if content[0] == "TITLE":
            info["Structure Title"].append(content[1])
        if content[0] == "COMPND" and content[1].startswith("MOLECULE: "):
            info["Name"].append(content[1].split("MOLECULE: ")[1])
        if content[0] == "COMPND" and content[1].startswith("CHAIN: "):
            info["Chains"].append(content[1].split("CHAIN: ")[1][0])
        if content[0] == "HET":
            lig = list(filter(lambda x: x != "", content[1].split(" ")))
            lig[-1] = int(lig[-1])
            info["Ligands"].append(lig)
```
The answer is that you can change them to `elif` as well, but there isn't really any need to do so, because there isn't any case for which more than one condition is true, so the result of consecutive `if` clauses, and an `if` clause followed by `elif`s will be the same.